### PR TITLE
Multiscale

### PR DESCRIPTION
**Pull request by alessiofumagalli**
So, apart from the contribution on the specific folder (paper/multiscale) there are few file changes that you should focus:
- src/porepy/fracs/mortars.py
- src/porepy/grids/grid_bucket.py
- src/porepy/grids/refinement.py
for the first two, now when you update the grids after a refinement/coarsening the internal structure of the grid bucket does not change. Meaning that the loop flow are the same. In the previous version a node with node_num small could have been move at the end. This gives the possibility to update all the new grids at the same time (see the mortars.py) and not just one by one.
for the last (refinement.py) the tags where not updated correctly.

Hopefully, a pre-print will appear soon. The actual merge will be done just after, but I prefer to start the discussion now.

### CODE REVIEW

**Comment by alessiofumagalli** [comment]
@IvarStefansson , why are you afraid? :P

**Comment by IvarStefansson** [comment]
"[U]sed for politely telling someone something that might make them sad, disappointed, or angry, for example when you cannot do what they want or do not agree with them" 
and
"Notice that this phrase doesn't actually mean that you're frightened; it's just a polite way to show that what you're going to say will be disappointing to the listener."

I'm also (actually) afraid that Eirik will not approve of our off-topic use of this pull request comment thread, as he probably receives e-mail for each comment...

File: src/porepy/fracs/mortars.py [hunk 1]
```diff
@@ -340,10 +340,11 @@ def replace_grids_in_bucket(gb, g_map={}, mg_map={}, tol=1e-6):
     for mg_old, mg_new in mg_map.items():
         update_mortar_grid(mg_old, mg_new, tol)
 
+    if g_map:
```
Comment (keileg) [inline src/porepy/fracs/mortars.py]: Please explain in a comment.

File: src/porepy/grids/grid_bucket.py [hunk 2]
```diff
@@ -582,21 +582,17 @@ def remove_nodes(self, cond):
...
+    def update_nodes(self, mapping):
         """
-        Update the grids giving the old and new values. The edges are updated
+        Update the grids giving old and new values. The edges are updated
         accordingly.
 
         Parameters:
-            new: List of the new grids. Can be a single element
-            old: List of the old grids, Can be a single element
+            mapping: A dictionary with the old labels as keys and new
```
Comment (keileg) [inline src/porepy/grids/grid_bucket.py]: By label, you mean grid?

File: src/porepy/grids/refinement.py [hunk 3]
```diff
@@ -243,28 +245,18 @@ def remesh_1d(g_old, num_nodes, tol=1e-6):
...
-            new_frac_face.append(nfi[0])
-
-    # This can probably be made more elegant
-    g.tags["fracture_faces"][new_frac_face] = True
-
-    # Fracture tips should be on the boundary only.
-    if np.any(g_old.tags["tip_faces"]):
-        g.tags["tip_faces"] = g.tags["domain_boundary_faces"]
+    # normally the tags are given at faces/point that are fixed the 1d mesh
+    # we use this assumption to proceed.
```
Comment (keileg) [inline src/porepy/grids/refinement.py]: Should this assumption be documented?

File: src/porepy/grids/refinement.py [hunk 4]
```diff
@@ -243,28 +245,18 @@ def remesh_1d(g_old, num_nodes, tol=1e-6):
...
+    for f_old in np.arange(g_old.num_faces):
```
Comment (keileg) [inline src/porepy/grids/refinement.py]: Please explain logic in for-loop, I can't read it .

File: src/porepy/numerics/fem/rt0.py [hunk 5]
```diff
@@ -12,18 +12,18 @@
 
 import porepy as pp
```
Comment (keileg) [inline src/porepy/numerics/fem/rt0.py]: At some point, we should replace the long imports with simply pp.foo for all methods and classes that are exposed at the top level. But not today.

File: src/porepy/numerics/fem/rt0.py [hunk 6]
```diff
@@ -331,6 +313,62 @@ def extract_p(self, g, up):
 
     # ------------------------------------------------------------------------------#
 
+    def project_u(self, g, u, data):
```
Comment (keileg) [inline src/porepy/numerics/fem/rt0.py]: This should be mentioned in a release note for this update

File: src/porepy/numerics/fem/rt0.py [hunk 7]
```diff
@@ -363,4 +401,14 @@ def massHdiv(self, K, c_volume, coord, sign, dim, HB):
         return np.dot(C.T, np.dot(N.T, np.dot(HB, np.dot(inv_K, np.dot(N, C)))))
 
 
+# ------------------------------------------------------------------------------#
+
+    def opposite_side_node(self, face_nodes, nodes, faces_loc):
```
Comment (keileg) [inline src/porepy/numerics/fem/rt0.py]: 1. Document function, I have no idea what it does.
2. Is this really much more general, and should be put somewhere related to grids?

**Comment by keileg** [comment]
This should be made into a new release. Either bump the version number yourself (setup.py and src/porepy/__init__.py), or notify me when you are done with the changes.

File: src/porepy/grids/refinement.py (hunk 3 above)
Comment (alessiofumagalli) [reply src/porepy/grids/refinement.py]: what do you propose?

File: src/porepy/numerics/fem/rt0.py (hunk 7 above)
Comment (alessiofumagalli) [reply src/porepy/numerics/fem/rt0.py]: it relay on the fact that the grid is made of simplex, so not so general

File: src/porepy/grids/refinement.py (hunk 3 above)
Comment (keileg) [reply src/porepy/grids/refinement.py]: Documentation in the function header? If it really is a basic assumption, it should not be tacit

File: src/porepy/numerics/fem/rt0.py (hunk 6 above)
Comment (alessiofumagalli) [reply src/porepy/numerics/fem/rt0.py]: It does, it's actually a great idea!

File: src/porepy/grids/refinement.py (hunk 3 above)
Comment (alessiofumagalli) [reply src/porepy/grids/refinement.py]: OK, I'll do it.

**Comment by keileg** [comment]
@alessiofumagalli Also see comments from codacy.

**Comment by alessiofumagalli** [comment]
OK, I took care of the comments (most of them) of codacy. Once the arvix is ready we can do the merge.

**Comment by keileg** [comment]
Please don't merge until the version number is bumped. I will take care of this - it needs some coordination with the branch robin_bc. @rbe051 please take note as well.

**Pull request merged**